keps/sig-autoscaling/draft-20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,170 @@
+---
+kep-number: draft-20181106
+title: In-place Update of Pod Resources
+authors:
+  - "@kgolab"
+  - "@bskiba"
+  - "@schylek"
+owning-sig: sig-autoscaling
+participating-sigs:
+  - sig-node
+  - sig-scheduling
+reviewers:
+  - "@bsalamat"
+  - "@derekwaynecarr"
+  - "@dchen1107"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-11-06
+last-updated: 2018-11-06
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# In-place Update of Pod Resources
+
+## Table of Contents
+
+   * [In-place Update of Pod Resources](#in-place-update-of-pod-resources)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [API Changes](#api-changes)
+         * [Flow Control](#flow-control)
+         * [Notes](#notes)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+      * [Alternatives](#alternatives)
+
+## Summary
+
+This proposal aims at allowing Pod resource requests & limits to be updated
+in-place, without a need to restart the Pod or its Containers.
+
+The **core idea** behind the proposal is to make PodSpec mutable with regards to
```

Comment:
The core idea make sense to me, and align with design in our side. We will list more details in comment soon. cc @ScorpioCPH 

One concern is the api change below seems "huge". Will we consider use annotations for some point?

Reply:
Our ref implementation (a bit out-of-date now https://github.com/Huawei-PaaS/kubernetes/pull/37/files) is based on annotations for communicating state between k components. However, one of the concerns we heard on our design doc was that annotations can allow affecting execution in undesirable ways.

This merged implementation @kgolab proposed is much cleaner. We should be able to do without annotations, with scheduler updating PodStatus.ContainerStatus.ResourceAllocated object, and kubelet consuming its values to invoke UpdateContainerResources assuming pod-fits-resources predicate says yes.

Reply:
Hi, here are some requirements from our customers:
- We want to resize resource not just on CPU, but also on memory and local ephemeral storage.
- Don't break the original API design (no changes in PodSpec) as there need to upgrade their system, which will be a huge work.
- If the node can't satisfy the increased resource requirements, throw an error and let user define what to do next (don't assign the pod to another available node automatically).

Reply:
So put these additional infos in annotations will be friendly to customers :)

Reply:
@resouer , @vinaykul - as far as I know annotations are discouraged from being an API surface, especially for requesting or controlling changes made by other components.

@ScorpioCPH - I don't see a problem though with using annotations to **report** actions/status in a human-readable way, aside from machine-readable conditions.

Reply:
> So put these additional infos in annotations will be friendly to customers :)

@ScorpioCPH We found event stream to be a good way to report last operation status. Thoughts?

Reply:
@ScorpoiCPH Please review updated design and comment if there are any concerns, hopefully your requirements have been addressed. I've tracked ephemeral storage resize as a future improvement in the KEP. I'll resolve this issue as many of the concerns have been addressed AFAIK.